std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+ * functions without relying on external build tools is impossible.
+ *
+ * The method is simple: Overallocate, manually align, and store the offset
+ * to the original behind the returned pointer.
+ *
+ * Align must be a power of 2 and 8 <= align <= 128.
+ */
+private void* XXH_alignedMalloc(size_t s, size_t align_) @trusted nothrow @nogc
+{
+    import core.stdc.stdlib : malloc;
+
+    assert(align_ <= 128 && align_ >= 8, "align out of range"); /* range check */
+    assert((align_ & (align_ - 1)) == 0, "(align_ & (align_ - 1)) != 0"); /* power of 2 */
+    assert(s != 0 && s < (s + align_), "s == 0 || s >= (s + align_)"); /* empty/overflow */
+    { /* Overallocate to make room for manual realignment and an offset byte */
+        xxh_u8* base = cast(xxh_u8*) malloc(s + align_);
+        if (base != null)
+        {
+            /*
+             * Get the offset needed to align this pointer.
+             *
+             * Even if the returned pointer is aligned, there will always be
+             * at least one byte to store the offset to the original pointer.
+             */
+            size_t offset = align_ - (cast(size_t) base & (align_ - 1)); /* base % align */
+            /* Add the offset for the now-aligned pointer */
+            xxh_u8* ptr = base + offset;
+
+            assert(cast(size_t) ptr % align_ == 0, "ptr % align_ != 0");
+
+            /* Store the offset immediately before the returned pointer. */
+            ptr[-1] = cast(xxh_u8) offset;
+            return ptr;
+        }
+        return null;
+    }
+}
+/*
+ * Frees an aligned pointer allocated by XXH_alignedMalloc(). Don't pass
+ * normal malloc'd pointers, XXH_alignedMalloc has a specific data layout.
+ */
+private void XXH_alignedFree(void* p) @trusted nothrow @nogc
+{
+    import core.stdc.stdlib : free;
+
+    if (p != null)
+    {
+        xxh_u8* ptr = cast(xxh_u8*) p;
+        /* Get the offset byte we added in XXH_malloc. */
+        xxh_u8 offset = ptr[-1];
+        /* Free the original malloc'd pointer */
+        xxh_u8* base = ptr - offset;
+        free(base);
+    }
+}
+
+private void XXH3_INITSTATE(XXH3_state_t* XXH3_state_ptr) @trusted nothrow @nogc
+{
+    (XXH3_state_ptr).seed = 0;
+}
+
+/*! @ingroup XXH3_family */
+XXH3_state_t* XXH3_createState() @trusted nothrow @nogc
+{
+    XXH3_state_t* state = cast(XXH3_state_t*) XXH_alignedMalloc((XXH3_state_t).sizeof, 64);
+    if (state == null)
+        return null;
+    XXH3_INITSTATE(state);
+    return state;
+}
+
+/*! @ingroup XXH3_family */
+XXH_errorcode XXH3_freeState(XXH3_state_t* statePtr) @trusted nothrow @nogc
+{
+    XXH_alignedFree(statePtr);
+    return XXH_errorcode.XXH_OK;
+}
+
+void XXH3_copyState(XXH3_state_t* dst_state, const XXH3_state_t* src_state) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memcpy;
+
+    memcpy(dst_state, src_state, (*dst_state).sizeof);
+}
+
+private void XXH3_reset_internal(XXH3_state_t* statePtr, XXH64_hash_t seed,
```

Comment:
This seems to be something similar to `.init` with some extra assignments.

Reply:
Yes, changes code and eleminated memset()